### PR TITLE
Set startPos at EOF in jsdoc token scanner so node end positions for nodes terminated at EoF are right

### PR DESCRIPTION
**Pull request by weswigham**
Fixes an incremental parsing crash @amcasey reported.

Note that `EoF` for jsdoc comments _also_ includes end of comment - our jsdoc parsing tests just don't include incremental validations.

### CODE REVIEW

**Comment by mhegazy** [comment]
@sandersn can you take a look

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -6471,7 +6471,23 @@ namespace ts {
                     return finishNode(result, end);
                 }
 
+                function isNextNonwhitespaceTokenEndOfFile(): boolean {
+                    nextJSDocToken();
+                    if (token() === SyntaxKind.EndOfFileToken) {
+                        return true;
+                    }
+                    if (token() === SyntaxKind.WhitespaceTrivia || token() === SyntaxKind.NewLineTrivia) {
+                        return lookAhead(isNextNonwhitespaceTokenEndOfFile); // We must use infinte lookahead, as there could be any number of newlines :(
```
Comment (sandersn) [inline src/compiler/parser.ts]: typo:infinite

Comment (sandersn) [reply src/compiler/parser.ts]: why do we need lookahead inside of lookahead? Is there a way to get around that?

File: src/services/classifier.ts [hunk 2]
```diff
@@ -706,16 +706,17 @@ namespace ts {
...
                             processElement((<JSDocTypeTag>tag).typeExpression);
+                            pos = tag.end;
                             break;
                         case SyntaxKind.JSDocReturnTag:
                             processElement((<JSDocReturnTag>tag).typeExpression);
+                            pos = tag.end;
                             break;
                     }
-
-                    pos = tag.end;
```
Comment (sandersn) [inline src/services/classifier.ts]: what prompts this change?

**Comment by mhegazy** [comment]
the node6 jest issue is still happening..

File: src/compiler/parser.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/parser.ts]: Yeah, I think you should be able to `return isNextNonwhitespaceTokenEndOfFile()` instead

File: tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/// <reference path="fourslash.ts" />
+
+// @noLib: true
+////function camelcase(flag) {
```
Comment (sandersn) [inline tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts]: lol irony in naming

File: src/compiler/parser.ts (hunk 1 above)
Comment (billti) [reply src/compiler/parser.ts]: Even just recursion for each line/whitespace seems suboptimal, but preferable, and it not like you'll have thousands of blank lines in a source file (hopefully!)

File: tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.noReturnType.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
         "0": {
             "kind": "JSDocReturnTag",
             "pos": 8,
-            "end": 16,
+            "end": 15,
```
Comment (billti) [inline tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.noReturnType.json]: Odd to see all these end position shift sometimes +2, sometime +1, and sometimes -1. I assume you've done a spot check and the new positions are (hopefully more) correct in each case?

File: src/compiler/parser.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/parser.ts]: Yeah. It's just a little unfortunate that _jsdoc parsing_ needs infinite lookahead - previously only arrow functions needed that.

File: tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+//// });
+////}
+////
+/////**
+//// * Pad `str` to `width`.
+//// *
+//// * @param {String} str
+//// * @param {Number} wid/*1*/
+goTo.marker('1');
+edit.insert("th\n@");
```
Comment (billti) [inline tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts]: Is there anything worth verifying here beyond "it doesn't crash"? I don't see it checking any lengths have been adjusted right (per the name).

Comment (weswigham) [reply tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts]: Just the not crashing. The position being wrong causes crashing.

File: tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.noReturnType.json (hunk 4 above)
Comment (weswigham) [reply tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.noReturnType.json]: Yeah, the end just no longer includes the trailing whitespace at the end of the comment.

File: tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts (hunk 5 above)
Comment (billti) [reply tests/cases/fourslash/incrementalJsDocAdjustsLengthsRight.ts]: I get that was the issue being addressed. I'm just suggesting that if you're going to add a fourslash test (which aren't cheap to run), might as well make it more useful and actually verify the JsDoc tags/spans also. It's a pretty narrow/specific test as-is that adds little value once this specific bug is fixed.

File: src/services/classifier.ts (hunk 2 above)
Comment (weswigham) [reply src/services/classifier.ts]: But... why are we doing that indentation analysis with token scanning and not with string manipulation on the comment text? We already manipulate it to strip leading and trailing newlines.

**Comment by weswigham** [comment]
@sandersn You may wanna double check the most recent change - I had to do some ugly stuff to keep `typdef` end positions in the right place because `parseTagComments` consumes tokens even if it doesn't capture anything. I suppose I could also just wrap the `parseTagComments` in a `tryParse` to reset the scanner if it fails, but that's not great, either.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -6848,10 +6849,12 @@ namespace ts {
                             typedefTag.typeExpression = childTypeTag && childTypeTag.typeExpression && !isObjectOrObjectArrayTypeReference(childTypeTag.typeExpression.type) ?
                                 childTypeTag.typeExpression :
                                 finishNode(jsdocTypeLiteral);
+                                end = jsdocTypeLiteral.end;
```
Comment (sandersn) [inline src/compiler/parser.ts]: is the indent on this statement correct? It looks weird to me.

File: src/compiler/parser.ts [hunk 7]
```diff
@@ -6820,6 +6820,7 @@ namespace ts {
                     typedefTag.comment = parseTagComments(indent);
 
                     typedefTag.typeExpression = typeExpression;
+                    let end: number;
```
Comment (sandersn) [inline src/compiler/parser.ts]: does this quirk apply to callback tags too?

Comment (sandersn) [reply src/compiler/parser.ts]: From our discussion: yes, and nested `@param` types too, but the problem isn't bad enough to fix right now. Instead we should fix it by fixing the jsdoc scanner to ignore whitespace (and change the way we generate tag descriptions).

**Pull request merged**